execution/executor/src/components/in_memory_state_calculator_v2.rs

```diff
@@ -110,56 +92,104 @@ impl InMemoryStateCalculatorV2 {
 
         let state_updates_vec = Self::get_sharded_state_updates(to_keep);
-        let updates: ShardedStateUpdates = Self::calculate_block_state_updates(&state_updates_vec);
-        let latest_checkpoint = base.current.clone();
-        let usage =
-            Self::calculate_usage(latest_checkpoint.usage(), &sharded_state_cache, &updates);
+        let last_checkpoint_index = to_keep
+            .iter()
+            .enumerate()
+            .rev()
+            .find(|(_, (txn, txn_output))| Self::need_checkpoint(txn, txn_output))
+            .map(|(index, _)| index);
 
+        let (updates_before_last_checkpoint, updates_after_last_checkpoint) =
+            if let Some(index) = last_checkpoint_index {
+                (
+                    Self::calculate_updates(&state_updates_vec[..=index]),
+                    Self::calculate_updates(&state_updates_vec[index + 1..]),
+                )
+            } else {
+                (
+                    create_empty_sharded_state_updates(),
+                    Self::calculate_updates(&state_updates_vec),
+                )
+            };
+
+        let num_txns = to_keep.len();
         let next_epoch_state = if new_epoch {
-            Some(Self::get_epoch_state(&sharded_state_cache, &updates)?)
+            ensure!(
+                last_checkpoint_index == Some(num_txns - 1),
+                "The last txn must be a reconfig for epoch change."
+            );
+            Some(Self::get_epoch_state(
+                &sharded_state_cache,
+                &updates_before_last_checkpoint,
+            )?)
         } else {
             None
         };
 
-        let (new_checkpoint, new_checkpoint_version) = {
-            let _timer = APTOS_EXECUTOR_OTHER_TIMERS_SECONDS
-                .with_label_values(&["make_checkpoint"])
-                .start_timer();
-            let latest_checkpoint_version = base.current_version;
-            let new_checkpoint_version =
-                Some(latest_checkpoint_version.map_or(0, |v| v + 1) + num_txns as u64 - 1);
-            let new_checkpoint = Self::make_checkpoint(
-                latest_checkpoint,
-                &updates,
-                usage,
-                ProofReader::new(proofs),
-            )?;
-            (new_checkpoint, new_checkpoint_version)
+        let proof_reader = ProofReader::new(proofs);
+        let latest_checkpoint = if last_checkpoint_index.is_some() {
+            Self::make_checkpoint(
+                base.current.clone(),
+                &updates_before_last_checkpoint,
+                &sharded_state_cache,
+                &proof_reader,
+            )?
+        } else {
+            base.base.clone()
         };
 
+        let first_version = base.current_version.map_or(0, |v| v + 1);
+        let mut latest_checkpoint_version = base.base_version;
+        let mut state_checkpoint_hashes = Vec::with_capacity(num_txns);
+        state_checkpoint_hashes.resize(num_txns, None);
```

Comment:
`vec![None; num_txns]` should work? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021

Reply:
Done